src/mpi/coll/ineighbor_alltoall/ineighbor_alltoall_allcomm_sched_linear.c

```diff
@@ -42,11 +42,20 @@ int MPIR_Ineighbor_alltoall_allcomm_sched_linear(const void *sendbuf, MPI_Aint s
         MPIR_ERR_CHECK(mpi_errno);
     }
 
-    /* receive needs to happen in the opposite order of sends.  This
-     * is to cover the case of Cartesian graphs where the same process
-     * might be both the left and right neighbor.  In this case, we
-     * need to make sure the first message (which is from the right
-     * neighbor) is received in the last buffer. */
+    /* Cartetian graph will result in multiple edges going to the same process when it is
+     * periodic and dim is 1 or 2, in which case, both left and right neighbor points to
+     * self or the other process. When that occurs, we are sending two messages to the
+     * same target and receiving two messages from the same target. Both messages are using
+     * the same tag, thus we need reverse the order of recvs to ensure correct matching.
+     * In the example of 1-dim cartesian graph, if we send in the order of left and right,
+     * the target need receive in the order of right and left to receive the correct messages.
+     *
+     * For all other case, the source and destination ranks should differentiate the messages
+     * and the order doesn't matter.
+     *
+     * The matching will be undefined if a general graph contains duplicate edges with
+     * unspecified orders.
```

Comment:
Is this comment talking about the MPI_Graph_create, et. al functions? Duplicate edges are not possible in those functions.

Reply:
The algorithm here is without the context of how the graph was created. Since the fix here is concerned with duplicated edges, I feel necessary to explain that it only accommodates a particular case of duplicated edges that is resulted from MPI_Cart_create. Let me rewrite this paragraph to be more specific.